### ISSUE TITLE
Proposal: Allow dynamic themes via JSON templates

### ISSUE THREAD

**Issue by pbarnum** [fyne-io/fyne#211]
It would be nice to be able to load and run custom themes during runtime.
* The app's main theme must still be written in Go and compiled with the app.
* New themes must follow strict guidelines in order to be a valid and usable theme.
* A app's theme should run after successfully loading.
* The app can handle multiple themes
  * This would require a menu system to switch between themes. Though this might be needed anyway if a user wanted to switch back to default after loading a custom theme.

**Comment by andydotxyz** [comment]
I like this proposal.
Theme loading is probably out of scope for the 1.1 release but if someone contributed some solid code we would probably accept it.

Part of the conversation may be about the actual storage format. JSON and XML are easy from a Go point of view but are they right for people who will edit the themes?

It may also relate to the [settings discussion](https://github.com/fyne-io/fyne/wiki/Settings-API).

**Comment by pbarnum** [comment]
Good point. I agree that JSON and XML are easy to work with and provide a simple and familiar interface, but it could get out of hand depending on a few factors:
* Number of customizations
* Level of detail within customizations
* Relationships between customizations
* Versioning
Perhaps a look into how other applications handle custom theming could answer some questions. VSCode comes to mind as does Sublime Text.

**Comment by asciifaceman** [comment]
VSCode uses json defined overrides of sane defaults, likely with a right to left merge somewhere in the line

**Comment by andydotxyz** [comment]
Depending on whether or not we change how themes work I don't think this needs to be multiple files. 
Unless you include the icon assets too...

**Comment by pbarnum** [comment]
That gets into `assets/` territory where these should be loaded separately IMO. Assets could follow a set scheme declared in the code, something like `assets/framework/right_border.png` or `assets/icons/home.png`

**Comment by andydotxyz** [comment]
Given that we have started using TOML for the FyneApp.toml build metadata, I wonder if we should do the same here.
It could be the same file even, but that seems less wise as an app may want to switch themes.
